### PR TITLE
Expose feedback cache API for Feedback 2.0 UX

### PR DESCRIPTION
**Pull request by JunDai**
### Description

The Feedback 2.0 flow allows to add more information on arrival. To support the new feedback UX change, the Core SDK should provide these APIs:
* cache user feedback
* retrieve cached feedback list
* post cached feedbacks

### Changelog

```
<changelog>Expose APIs to support caching user feedback events.</changelog>
```

### CODE REVIEW

**Comment by JunDai** [comment]
@korshaknn / @Guardiola31337 : this is just the proposal of feedback caching change.
please help to review if this makes sense before I integrate it with the UI SDK.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:684]: why do we need both `cacheUserFeedback ()` and `postUserFeedback` ?
what usecase do we have for each of them?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:712]: looks like we have the next flow
- cache events with `MapboxNavigationTelemetry`
- take them and add some details (what class is responsible for such changes?)
- use `MapboxNavigationTelemetry` to post events

is it a good idea to take events out of `MapboxNavigationTelemetry` to change them?
what happens if user closes the app/ cancel the session? with current implementation all pending events will be sent immediately. But looks like cached events will be lost

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:115]: does the order matters?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:214]: why  do we need to update old events with new from a `cachedFeedbackEventList `?
can we just post new events because they are already updated?

**Review by korshaknn — COMMENTED** [review]
we have some tailwork with Telemetry, it would be good to finish and test it first. And after that change/add something new.
and also we need to cover each new method with a bunch of test

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:684]: in short, feedback 2.0 flow has different concept of 1tap and 2tap flow.
* for 1tap flow, user will use the `postUserFeedback` so the feedback will be queued to send.
* for 2tap flow, user will use the `cacheUserFeedback` so the feedback is cached and will be able to do update to it in a later time.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:712]: that's correct. 
* the flow you described is correct. That's why you can see I expose a new `CachedNavigationFeedbackEvent` instead of the internal `NavigationFeedbackEvent` and also only allow the `feedbackSubType` and `description` to be modified.
* yes, we take the risk of losing those cached events base on agreement of the feedback 2.0 flow discussion. I mentioned you in that ticket for reference.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:115]: yes, we want to show the cached feedback events in the order of when it's created.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:214]: when we post new events, we lose the location information. we only need the user's input of `feedbackSubType` and `description`, that's why I do this way.

**Comment by JunDai** [comment]
> we have some tailwork with Telemetry, it would be good to finish and test it first. And after that change/add something new.
> and also we need to cover each new method with a bunch of test

yes, it completely makes sense. This draft is to give core team an idea about the proposal and make sure we can get an agreement of the new API. Also I will add test once the APIs are decided.

**Comment by JunDai** [comment]
@korshaknn - when is the tailwork of Telemetry going to finish? and what do you think about the proposed apis?

**Comment by korshaknn** [comment]
@JunDai thanks for your explanations! your changes LGTM! We still have some issues with telemetry, some of them might be on NN side. But I think it's not a blocker for your PR.
cc @Guardiola31337

**Comment by JunDai** [comment]
@korshaknn / @Guardiola31337 : this PR is ready for reviewing, can you please take a look?
This is part of the Feedback 2.0 implementation. I would create another PR for the UI part so we can focus on our review on different areas.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:171]: let's move lambda to the last params position

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:202]: return value for readability

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (korshaknn) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:149]: we have a `private fun reset()` to reset data on `init` and on `sessionStop`.
do we need to move `feedbackEventCacheMap.clear()` to `reset()` ?
or it shouldn't be cleaned until new `initialize ` / `postCachedUserFeedback `?

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt
Comment (korshaknn) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:350]: not needed

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt
Comment (korshaknn) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:357]: not needed

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt
Comment (korshaknn) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:407]: just to clarify
right now `cancel` event is always a final event in each session.
it means that on `cancel` event we call `locationsCollector.flushBuffers` and we immediately post all pending events (which are waiting for location buffer collection).

so, cachedEvents might be sent `after` `cancel`. 
@Guardiola31337

**Review by korshaknn — COMMENTED** [review]
added some comments

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:149]: > it shouldn't be cleaned until new initialize / postCachedUserFeedback 

that's correct.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt
Comment (JunDai) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:357]: if we don't add this line, I think we are not flushing the location buffer which in result will not call the cached user feedback event's `onBufferFull` callback.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt
Comment (JunDai) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:407]: correct.

**Comment by JunDai** [comment]
@korshaknn / @Guardiola31337 - PR updated to address comments from latest review.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt
Comment (korshaknn) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:352]: got it, you are right!

**Review by korshaknn — APPROVED** [review]
LGTM! 
@Guardiola31337 please take a look

File: libnavigation-core/api/current.txt
Comment (Guardiola31337) [inline libnavigation-core/api/current.txt:16]: This is breaking SemVer. Notice `String![]? feedbackSubType = emptyArray()` vs `String![] feedbackSubType = emptyArray()`.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:677]: See comment above re: breaking SemVer

**Comment by JunDai** [comment]
@Guardiola31337 - would you mind to check again?
* I added one more test case `only_internal_cached_feedback_sent` to address the issue that we discussed offline.
* also reverted the change which causes the `SEMVER_MAJOR` concern.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:690]: We should be explicitly and add a note stating that this is the case only when in _Active Guidance_ (Feedback events are only handled in that mode) and as long as `postCachedUserFeedback` is called before transitioning between modes as that automatically will flush all the pending events. If not, all those events will be lost (as they're not push, only cached). BTW is that the behavior we want for such a use case? Have we tested this scenario? Wondering if we could be smarter when it comes to losing events 🤔 

Refs. https://github.com/mapbox/mapbox-navigation-android/pull/3724#discussion_r515205044

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:220]:
```suggestion
        log("post cached user feedback events")
```

**Review by Guardiola31337 — COMMENTED** [review]
Thanks for addressing the feedback @JunDai 

Left a couple of minor comments / questions to discuss before merging here.

**Comment by JunDai** [comment]
@Guardiola31337 - per @asinghal22 , we don't need to expose these APIs to external customer. so as we discussed offline, I moved them into an internal package. Also added more documents to those APIs per your suggestion. Please take a look again. Thanks!

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:12]:
```suggestion
 * This object is only a wrapper of [MapboxNavigationTelemetry] to cache/retrieve/post a navigation.feedback event.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:16]:
```suggestion
 * 1. [cacheUserFeedback] to generate a navigation.feedback event in [MapboxNavigationTelemetry] internally.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:18]:
```suggestion
 * 3. make possible changes to the list that is returned by [getCachedUserFeedback].
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:89]: What about adding the note from above here too?

```kotlin
 * Note:
 * 1. Only the feedback events created by [cacheUserFeedback] and returned by [getCachedUserFeedback]
 * are valid in [postCachedUserFeedback]. All user manually created events will be dropped.
 * 2. The cache APIs should only be used when in Active Guidance mode since navigation.feedback events are only handled in that mode.
 * 3. [postCachedUserFeedback] should be called before transitioning between modes. Otherwise, all cached events will be lost.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/CachedNavigationFeedbackEvent.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/CachedNavigationFeedbackEvent.kt:13]: `CachedNavigationFeedbackEvent` can be `internal` or moved into the `internal` package otherwise.

**Review by Guardiola31337 — COMMENTED** [review]
Left some minor comments @JunDai

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/CachedNavigationFeedbackEvent.kt
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/CachedNavigationFeedbackEvent.kt:13]: this event should be exposed, because the UI side will need to use it. Please refer to the Feedback UX PR.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/CachedNavigationFeedbackEvent.kt:13]: It can be "public" but under the `internal` package so it'll be accessible from UI side.

> or moved into the `internal` package

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt:15]: I believe this can be now a `data class` instead of a `Builder`. BTW if we want to leave it as is, why are we missing the `toString`? cc @LukasPaczos

**Review by Guardiola31337 — APPROVED** [review]
Thanks for addressing the feedback @JunDai :shipit:

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt]: `data class` exposes a public `.copy()` method. If someone relies on that method, their build will break when we add a new constructor parameter, even if the constructor itself is `internal`.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt]: The class itself is `internal` so that shouldn't be a problem, but as I mentioned I'm fine with `Builder` too, as long as we add the `toString` 😅

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt]: Oh I didn't notice you moved it to `internal` package :+1: if it was public, it wouldn't have to have a builder pattern anyway, just a `class CachedNavigationFeedbackEvent internal constructor` with `equals/hash/toString` overrides.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt]: Yeah, that's correct.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt]: I meant "builder" 😅

**Pull request merged**
